src/main/java/net/dv8tion/jda/core/entities/impl/JDAImpl.java

```diff
@@ -132,6 +135,13 @@ public void login(String token, ShardInfo shardInfo, SessionReconnectQueue recon
         if (token == null || token.isEmpty())
             throw new LoginException("Provided token was null or empty!");
 
+        if (shardInfo != null)
+        {
+            contextMap.put("jda.shard", shardInfo.getShardString());
+            contextMap.put("jda.shard.id", String.valueOf(shardInfo.getShardId()));
+            contextMap.put("jda.shard.total", String.valueOf(shardInfo.getShardTotal()));
+        }
+        MDC.setContextMap(contextMap);
```

Comment:
This nukes any MDC values that client code might have set 😢 
In your own threads I don't mind manually managing the contextMap, but this is rather destructive
Since buildAsync isn't enough, you'd have to invoke buildAsync asynchronously and explain why that's done that way and hope no one refactors it away

Reply:
The client is able to control the map

Reply:
I understand your concern though. I could make use of MDC.MDCClosable and just put all the values in the existing context for the build process

Reply:
This seems like a fair alternative
```java
// set MDC metadata for build thread
Map<String, String> previousContext = MDC.getCopyOfContextMap();
contextMap.forEach(MDC::put);
// init code
MDC.setContextMap(previousContext);
```

Reply:
As long as it's addressed I'm fine with it 👍